cli/main.go

```diff
@@ -191,50 +194,69 @@ func main() {
 	if opts.Config == "" {
 		fatal(errors.New("config path cannot be empty"))
 	}
-
 	configDir := opts.Config
-	ctx = config.WithDir(ctx, configDir)
+	config.WithDir(configDir)
 
 	currentContext := determineCurrentContext(opts.Context, configDir, opts.Hosts)
+	apicontext.WithCurrentContext(currentContext)
 
 	s, err := store.New(configDir)
 	if err != nil {
 		mobycli.Exec(root)
 	}
+	store.WithContextStore(s)
 
 	ctype := store.DefaultContextType
 	cc, _ := s.Get(currentContext)
 	if cc != nil {
 		ctype = cc.Type()
 	}
 
+	service, err := getBackend(ctype, configDir, opts)
+	if err != nil {
+		fatal(err)
+	}
+	backend.WithBackend(service)
+
 	root.AddCommand(
 		run.Command(ctype),
 		compose.Command(ctype),
 		volume.Command(ctype),
 	)
-	if ctype == store.DefaultContextType || ctype == store.LocalContextType {
-		cnxOptions := cliflags.CommonOptions{
-			Context:   opts.Context,
-			Debug:     opts.Debug,
-			Hosts:     opts.Hosts,
-			LogLevel:  opts.LogLevel,
-			TLS:       opts.TLS,
-			TLSVerify: opts.TLSVerify,
+
+	if err = root.ExecuteContext(ctx); err != nil {
+		handleError(ctx, err, ctype, currentContext, cc, root)
+	}
+	metrics.Track(ctype, os.Args[1:], metrics.SuccessStatus)
+}
+
+func getBackend(ctype string, configDir string, opts cliopts.GlobalOpts) (backend.Service, error) {
+	switch ctype {
+	case store.DefaultContextType, store.LocalContextType:
+		configFile, err := cliconfig.Load(configDir)
+		if err != nil {
+			return nil, err
+		}
+		options := cliflags.CommonOptions{
+			Context: opts.Context,
+			Debug:   opts.Debug,
+			Hosts:   opts.Hosts,
 		}
 
 		if opts.TLSVerify {
-			cnxOptions.TLSOptions = opts.TLSOptions
+			options.TLSOptions = opts.TLSOptions
 		}
-		ctx = apicontext.WithCliOptions(ctx, cnxOptions)
+		apiClient, err := command.NewAPIClientFromFlags(&options, configFile)
+		if err != nil {
+			return nil, err
+		}
+		return local.NewService(apiClient), nil
 	}
-	ctx = apicontext.WithCurrentContext(ctx, currentContext)
-	ctx = store.WithContextStore(ctx, s)
-
-	if err = root.ExecuteContext(ctx); err != nil {
-		handleError(ctx, err, ctype, currentContext, cc, root)
+	service, err := backend.Get(ctype)
+	if errdefs.IsNotFoundError(err) {
```

Comment:
is there a specific reason to return nil without error here ?

Reply:
or do we have cases returning a non-nil service with IsNotFoundError ??

Reply:
This happens for `context create` command
Backend is actually optional for a few commands